gufe/components/proteincomponent.py

```diff
@@ -287,24 +287,51 @@ def _get_ion_charge(ion_key):
         # Add Additionals
         # Formal Charge
         netcharge = 0
-        for a in rd_mol.GetAtoms():
-            atom_name = a.GetMonomerInfo().GetName().strip()
-            atomic_num = a.GetAtomicNum()
-
-            connectivity = sum(_BONDORDER_TO_ORDER[bond.GetBondType()] for bond in a.GetBonds())
-            default_valence = periodicTable.GetDefaultValence(atomic_num)
 
+        def _get_formal_charge(valence: int, connectivity: int, atom_name: str) -> int:
             if connectivity == 0:  # ions
                 fc = _get_ion_charge(atom_name)
-            elif default_valence > connectivity:  # negative charge
-                fc = -(default_valence - connectivity)
-            elif default_valence < connectivity:  # positive charge
-                fc = +(connectivity - default_valence)
+            elif valence > connectivity:  # negative charge
+                fc = -(valence - connectivity)
+            elif valence < connectivity:  # positive charge
+                fc = +(connectivity - valence)
             else:
-                fc = 0  # neutral
+                fc = 0
+            return fc
 
-            a.SetFormalCharge(fc)
-            a.UpdatePropertyCache(strict=True)
+        for atom_id, a in enumerate(rd_mol.GetAtoms()):  # TODO: should this index start at 1?
```

Comment:
re: indexing, rdkit throws the valence error indexing from 0, but the actual atom ids are indexed from 1?

Reply:
I remember running into this and yes you need to match the indexing in the PDB so it makes sense.

Reply:
done!